### DISCUSSION TITLE
How do I get users created 'FROM EXTERNAL PROVIDER'

### DISCUSSION THREAD

**Discussion by ClaudioESSilva** [microsoft/sqlmanagementobjects#113] [Q&A]
Hi,

When using `$db.Users` (which returns `[Microsoft.SqlServer.Management.Smo.User]`) I can't get the ones that were created using `EXTERNAL PROVIDER`

Ex:
``` sql
CREATE USER [blah@****.onmicrosoft.com] FROM EXTERNAL PROVIDER
```

My understanding is that it doesn't show because `issqluser` (SQL Server) property (from `sys.sysusers`) is `0` (False)
![image](https://user-images.githubusercontent.com/19521315/192897850-c700b149-c1a0-4d5c-a487-fe87841152d8.png)

From `sys.database_principals` it shows as `EXTERNAL_USER`
![image](https://user-images.githubusercontent.com/19521315/192898902-32bf5d7c-3e8b-4b2d-8db8-2c540e6360ae.png)

Does SMO expose this type of user? 
How can we get them using SMO?

**Comment by shueybubbles** [comment]
What type of server are you connected to? SMO is likely out of date for external users on SQL Managed Instance, since it only checks for external users on Azure DB. See https://github.com/microsoft/sqlmanagementobjects/blob/98093604e9ede97aa2a9b066279f0bd9d9557c4f/src/Microsoft/SqlServer/Management/SqlEnum/xml/user.xml#L77

Users are found using `sys.database_principals`

**Reply by ClaudioESSilva** [reply]
It's exactly an Azure SQL Managed Instance 🥲

**Reply by shueybubbles** [reply]
are the users missing from the collection completely, or just have the wrong type?

**Reply by ClaudioESSilva** [reply]
They are missing. Like they are being filtered by `Type = 'S'`

**Reply by shueybubbles** [reply]
This is the SMO query I see SSMS run when I expand Users node in an on-premises DB, which should be the same query SMO runs when querying users on MI:
```sql
SELECT
                                   u.name AS [Name],
                                   'Server[@Name=' + quotename(CAST(
                                           serverproperty(N'Servername')
                                          AS sysname),'''') + ']' + '/Database[@Name=' + quotename(db_name(),'''') + ']' + '/User[@Name=' + quotename(u.name,'''') + ']' AS [Urn],
                                   u.create_date AS [CreateDate],
                                   u.principal_id AS [ID],
                                   CAST(CASE dp.state WHEN N'G' THEN 1 WHEN 'W' THEN 1 ELSE 0 END AS bit) AS [HasDBAccess]
                                   FROM
                                   sys.database_principals AS u
                                   LEFT OUTER JOIN sys.database_permissions AS dp ON dp.grantee_principal_id = u.principal_id and dp.type = CO
                                   WHERE
                                   (u.type in ('U', 'S', 'G', 'C', 'K' ,'E', 'X'))
                                   ORDER BY
                                   [Name] ASC
```

**Reply by ClaudioESSilva** [reply]
Query got cropped on the `LEFT OUTER JOIN` line.
I think it's `COLLATE` something. Can you edit the snippet?

**Reply by shueybubbles** [reply]
`CO` is the `type` column value in `sys.database_permissions`
eg
```
class	class_desc	major_id	minor_id	grantee_principal_id	grantor_principal_id	type	permission_name	state	state_desc
0	DATABASE	0	0	1	1	CO  	CONNECT	G	GRANT
```

**Reply by ClaudioESSilva** [reply]
this bit is correct? 🤔 
![image](https://user-images.githubusercontent.com/19521315/193115825-9e50f69f-4027-4814-b288-b9f15986d302.png)

Should be `'CO'` now I get it!

**Reply by shueybubbles** [reply]
sorry, the output window in SSMS hides a lot of complexity. The real query ends up being dynamic SQL

```sql
exec sp_executesql N'SELECT
u.name AS [Name],
''Server[@Name='' + quotename(CAST(
        serverproperty(N''Servername'')
       AS sysname),'''''''') + '']'' + ''/Database[@Name='' + quotename(db_name(),'''''''') + '']'' + ''/User[@Name='' + quotename(u.name,'''''''') + '']'' AS [Urn],
u.create_date AS [CreateDate],
u.principal_id AS [ID],
CAST(CASE dp.state WHEN N''G'' THEN 1 WHEN ''W'' THEN 1 ELSE 0 END AS bit) AS [HasDBAccess]
FROM
sys.database_principals AS u
LEFT OUTER JOIN sys.database_permissions AS dp ON dp.grantee_principal_id = u.principal_id and dp.type = @_msparam_0
WHERE
(u.type in (''U'', ''S'', ''G'', ''C'', ''K'' ,''E'', ''X''))
ORDER BY
[Name] ASC',N'@_msparam_0 nvarchar(4000)',@_msparam_0=N'CO'
```

**Reply by ClaudioESSilva** [reply]
Thanks @shueybubbles !
The query returns the 'E' as expected.


## Found the problem
I can also see it from dbatools (SMO).
It's a cache/lack of refresh problem after adding the user (using TSQL and not refreshing the collection)

If I reconnect, to refresh my SMO.Server, object then it already appears.

Sorry about the confusion generated 😅 and thanks for the quick support!
